New Design for Package Search Dropdown

AndroGenius-codes opened the pull request:
Resolves https://github.com/ocaml/ocaml.org/issues/1602.

@sabine  This is a first draft for the search feature. i am going to attach a video of what i have done so far and I need some clarification and thoughts on the design.
1)do i need to change the size of the search bar in the header or should i leave it the way it is.
2) increasing the size of the search bar to match the width of the drop down will make the header too busy so i think we should either reduce the width of the dropdown to match the width of the search bar or let the width of the drop down be wider than that of the search bar.
3)I need your thoughts on the search bar on the package page. what do i need to change.
[search feature.webm](https://github.com/ocaml/ocaml.org/assets/70571088/7483d5cb-1eee-47a0-8d5a-98ca93358ea9)

N/B: i had to use vanilla css because tailwind was a bit limiting, i could not back changes to hover background, i would appreciate a little hint on how to overcome this.

sabine commented:
> 1)do i need to change the size of the search bar in the header or should i leave it the way it is.
> 2) increasing the size of the search bar to match the width of the drop down will make the header too busy so i think we should either reduce the width of the dropdown to match the width of the search bar or let the width of the drop down be wider than that of the search bar.

The width of the dropdown and the search bar in the header should stay the same, in this patch.

> 3)I need your thoughts on the search bar on the package page. what do i need to change.

* It looks like there is something off with the hover state because it spans both the package name and the docs link. These are separate links.
* The design has rounded corners on the hover state and the text color is white.
* The width of the dropdown on the `packages.eml` page does not look to be the same as the width of the search input.
* This pull request seems to include a commit that does not belong here: https://github.com/ocaml/ocaml.org/pull/1608/commits/d1529f870d89bcf439ea9604eca4cdf9501fe49f. The simplest way to remove this commit from the PR is to interactively rebase (e.g. `git rebase -i main`). If you try to do so, please make a backup copy of your branch first, in case something goes wrong.

AndroGenius-codes commented:
> > 1)do i need to change the size of the search bar in the header or should i leave it the way it is.
> > 2) increasing the size of the search bar to match the width of the drop down will make the header too busy so i think we should either reduce the width of the dropdown to match the width of the search bar or let the width of the drop down be wider than that of the search bar.
> 
> The width of the dropdown and the search bar in the header should stay the same, in this patch.
> 
> > 3)I need your thoughts on the search bar on the package page. what do i need to change.
> 
> * It looks like there is something off with the hover state because it spans both the package name and the docs link. These are separate links.
> * The design has rounded corners on the hover state and the text color is white.
> * The width of the dropdown on the `packages.eml` page does not look to be the same as the width of the search input.
> * This pull request seems to include a commit that does not belong here: [d1529f8](https://github.com/ocaml/ocaml.org/commit/d1529f870d89bcf439ea9604eca4cdf9501fe49f). The simplest way to remove this commit from the PR is to interactively rebase (e.g. `git rebase -i main`). If you try to do so, please make a backup copy of your branch first, in case something goes wrong.

Thank you @sabine. I will work on the changes

AndroGenius-codes commented:
@sabine i have fixed the ui issues, i also fixed a small 
[screen-feature-corrected.webm](https://github.com/ocaml/ocaml.org/assets/70571088/009499ec-c752-4282-a821-d2191ee53db3)
[responsive.webm](https://github.com/ocaml/ocaml.org/assets/70571088/16e5d98d-09cc-4619-a61d-510727f851f6)
responsive issue from the search bar on the package.eml pages

sabine commented on src/ocamlorg_frontend/components/header.eml:
```diff
@@ -23,8 +23,9 @@ let render
-  <span class="pl-2 font-semibold">Or go to:</span>
-  <a class="flex py-2 px-4 gap-4 hover:bg-primary-100 font-semibold hover:font-semibold text-primary-600" href="<%s Url.api %>">
+  <hr>
+  <span class="pl-2 font-normal">Or go to</span>
```
I have an idea here:

It probably makes sense to hide this span when we don't have any search results. What do you think?

AndroGenius-codes replied:
i tried to do this but it kept bring up some ui challenges because even with that span hidden the drop down background will always show. only way to implement this is to hide the dropdown div until at least one letter is typed on the input bar. unfortunately i was not able to do this

sabine replied:
the dropdown background showing should be fine because there's always the standard library link in it

AndroGenius-codes replied:
alright, i think my last commit has sorted this issue

sabine reviewed (commented):
Hey that's looking really good already! Here's some nitpicking, ideas and feedback.

AndroGenius-codes commented:
@sabine thank you so much for your feedback I'll start working on it immediately

AndroGenius-codes commented:
here are the new changes i was able to make 
[screen-capture (1).webm](https://github.com/ocaml/ocaml.org/assets/70571088/c253fe31-a0b9-4219-9c35-885f85845284)

AndroGenius-codes commented:
@sabine here are the changes from the last commit i made. i removed the "searching" indicator because it kept messing with position of other divs in the dom and it does not really add anything to the current search functionality. 
[screen-capture (2).webm](https://github.com/ocaml/ocaml.org/assets/70571088/4bd22c5c-7eb5-48e6-9faa-c199c039c0ad)

sabine commented on src/ocamlorg_frontend/css/styles.css:
```diff
@@ -255,3 +255,11 @@ details summary svg svg {
+.dropdown {
+  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
+}
+.dropdown-link:hover{
+  background-color:#00308F;
+  border-radius: 5px;
+}
```
This should be possible to do in tailwind with `hover:bg-[#00308F]` and choosing the closest available border radius from the tailwind documentation.

AndroGenius-codes replied:
Alright then. The nearest radius was 6px and it didn't give me what I wanted so I decided to go for 5px with vanilla css. I will fix that

sabine commented on src/ocamlorg_frontend/pages/packages_autocomplete_fragment.eml:
```diff
@@ -30,16 +30,15 @@ let render
-  </p>
+    <p class="break-words">
+      <%s (if total = List.length packages then "" else "Showing " ^ string_of_int (List.length packages) ^ " out of ")
+      ^ string_of_int total ^ (if total = 1 then "search results" else " packages") %>
```
```suggestion
      ^ string_of_int total ^ (if total = 1 then " package" else " packages") %>
```

sabine reviewed (commented):
I see a small alignment issue:

![Screenshot from 2023-10-17 18-51-07](https://github.com/ocaml/ocaml.org/assets/6594573/815967dd-c5c6-457d-94f3-2c8e1e65bfd7)

I'm not 100% sure about the width of the dropdown on the top right search bar. Will ask our designer what to do.

AndroGenius-codes commented:
> I see a small alignment issue:
> 
> ![Screenshot from 2023-10-17 18-51-07](https://user-images.githubusercontent.com/6594573/275915572-815967dd-c5c6-457d-94f3-2c8e1e65bfd7.png)
> 
> I'm not 100% sure about the width of the dropdown on the top right search bar. Will ask our designer what to do.
Alright, will be on standby to implement any changes . I personally think the drop down should take the width of the search bar. It's neater that way.

Clairevanden commented:
@AndroGenius-codes Hey, I am Claire the UX/UI designer. Thank you for the work, I just have 2 small commets . See screenshot. Thanks If any questions let me know. 
<img width="2283" alt="Screenshot 2023-10-18 at 12 22 22" src="https://github.com/ocaml/ocaml.org/assets/123180459/d9c96ee2-56ed-416b-b118-0718c2dc152b">

AndroGenius-codes commented:
> @AndroGenius-codes Hey, I am Claire the UX/UI designer. Thank you for the work, I just have 2 small commets . See screenshot. Thanks If any questions let me know. <img alt="Screenshot 2023-10-18 at 12 22 22" width="2000" src="https://user-images.githubusercontent.com/123180459/276212065-d9c96ee2-56ed-416b-b118-0718c2dc152b.png">

Thank you Claire. For the font size on hover do you want me to make them the same as the standard library link on hover or make it small as in see more link

AndroGenius-codes commented:
> @AndroGenius-codes Hey, I am Claire the UX/UI designer. Thank you for the work, I just have 2 small commets . See screenshot. Thanks If any questions let me know. <img alt="Screenshot 2023-10-18 at 12 22 22" width="2000" src="https://user-images.githubusercontent.com/123180459/276212065-d9c96ee2-56ed-416b-b118-0718c2dc152b.png">

Thank you Claire. I have made the changes

AndroGenius-codes commented:
Hi @sabine good morning, I think I am done with this task.

sabine commented on src/ocamlorg_frontend/pages/packages.eml:
```diff
@@ -6,7 +6,15 @@ Layout.render
 ~active_top_nav_item:Header.Packages @@
-<div class="intro-section-simple">
+
+let search_dropdown () =
+  <div id="packages-search-results" aria-live="polite" ></div>
+  <a class="flex py-2 px-4 gap-4 hover:bg-primary-100 font-normal rounded-md text-primary-600" href="<%s Url.api %>">
+    Standard Library API
+    <%s! Icons.arrow_top_right_on_square "w-6 h-6"%>
+  </a>
+in
```
seems to be the same function as in `header.eml`, should be factored out to a place where it can be shared.

`packages_autocomplete_fragment.eml` seems to be a reasonable place.

sabine commented on src/ocamlorg_frontend/pages/packages.eml:
```diff
@@ -6,7 +6,15 @@ Layout.render
+
+let search_dropdown () =
+  <div id="packages-search-results" aria-live="polite" ></div>
+  <a class="flex py-2 px-4 gap-4 hover:bg-primary-100 font-normal rounded-md text-primary-600" href="<%s Url.api %>">
```
```suggestion
  <a class="flex py-2 px-2 mx-2 gap-4 hover:bg-primary-100 font-normal rounded-md text-primary-600" href="<%s Url.api %>">
```

there's still something about the alignment, I think this makes the StdLib link align better with the rest when hovered.

sabine commented on src/ocamlorg_frontend/pages/packages_autocomplete_fragment.eml:
```diff
@@ -51,22 +50,25 @@ let render
         href="<%s Url.Package.overview package.name %>"
         :class='row == <%s string_of_int i %> && col == 0 ? "bg-search-keyboard-cursor text-white": ""'
-        class="flex-grow px-2 py-2 leading-6 font-semibold hover:bg-primary-100 hover:text-primary-700 inline-block"
+        class="flex-grow px-2 py-2 leading-6 font-normal hover:text-white hover:bg-search-result-background-blue rounded-md inline-block"
```
```suggestion
        class="flex-grow ml-2 px-2 py-2 leading-6 font-normal hover:text-white hover:bg-search-result-background-blue rounded-md inline-block"
```

sabine commented on src/ocamlorg_frontend/components/forms.eml:
```diff
@@ -1,6 +1,6 @@
 let search_input ?button_attrs ?input_attrs ?dropdown_html ~label ~name _class =
   let form_dropdown html = 
-    <div class="dropdown z-10 absolute rounded-md w-full top-1 mt-10 p-2 bg-default dark:bg-dark-default border border-primary-600">
+    <div class="dropdown w-full lg:w-[32rem] xl:w-[32rem] z-10 absolute rounded-md mt-12 shadow-2xl top-1 p-2 bg-default dark:bg-dark-default shadow-3xl ">
```
```suggestion
    <div class="dropdown w-full lg:w-[32rem] xl:w-[32rem] z-10 absolute rounded-md mt-12 shadow-2xl top-1 p-2 bg-default dark:bg-dark-default shadow-3xl">
```

sabine reviewed (commented):
Here's some minor adjustments in margin to match the design more closely.

`search_dropdown` function should be refactored to avoid repeating it.

sabine reviewed (approved):
Thanks for the contribution @AndroGenius-codes! Well done!

AndroGenius-codes commented:
> Thanks for the contribution @AndroGenius-codes! Well done!

Thank you sabine
